File: app/src/main/java/com/arny/celestiatools/ui/fragments/OrbitCalcFragment.java
package com.arny.celestiatools.ui.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import android.widget.EditText;
import android.widget.TextView;
import com.arny.arnylib.utils.CalculatorDialog;
import com.arny.arnylib.utils.DateTimeUtils;
import com.arny.arnylib.utils.MathUtils;
import com.arny.arnylib.utils.Utility;
import com.arny.celestiatools.R;
import com.arny.celestiatools.utils.astro.CircularMotion;
import com.arny.celestiatools.utils.astro.EllipseMotion;
import com.arny.celestiatools.utils.astro.OrbitCalcMotion;

/**
 * A simple {@link Fragment} subclass.
 */
public class OrbitCalcFragment extends Fragment {

	private Context context;
	private EditText edtMass;
	private EditText edtRadius;
	private EditText edtHeightPeri;
	private EditText edtVelPeri;
	private TextView tvResult;
	private double mass;
	private double radius;
	private double Hp;
	private double Vp;

	public OrbitCalcFragment() {
		// Required empty public constructor
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_orbit_calc, container, false);
		context = container.getContext();
		initUI(view);
		initListeners();
		return view;
	}

	private void initListeners() {
		edtMass.setOnClickListener(view -> new CalculatorDialog(context,edtMass.getText().toString(), result -> {
			edtMass.setText(result);
			calc();
		}).show());
		edtRadius.setOnClickListener(view -> new CalculatorDialog(context,edtRadius.getText().toString(), result -> {
			edtRadius.setText(result);
			calc();
		}).show());
		edtHeightPeri.setOnClickListener(view -> new CalculatorDialog(context,edtHeightPeri.getText().toString(), result -> {
			edtHeightPeri.setText(result);
			calc();
		}).show());
		edtVelPeri.setOnClickListener(view -> new CalculatorDialog(context,edtVelPeri.getText().toString(), result -> {
			edtVelPeri.setText(result);
			calc();
		}).show());
	}

	private void calc() {
		validateVariables();
		StringBuilder results = new StringBuilder();
		if (Vp <= 0) {
			CircularMotion circularMotion = OrbitCalcMotion.calcCircularOrbit(mass, radius, Hp);
			results.append("1я космическая:").append(MathUtils.round(circularMotion.getV1(), 6)).append("м/с\n");
			results.append("2я космическая:").append(MathUtils.round(circularMotion.getV2(), 6)).append("м/с\n");
			results.append("Ускорение:").append(MathUtils.round(circularMotion.getUskorenie(), 6)).append("м/с\n");
			int hour = circularMotion.getHour();
			int min = circularMotion.getMin();
			int sec = circularMotion.getSec();
			results.append("Период:").append(DateTimeUtils.convertTime(DateTimeUtils.convertTime(hour, min, sec)));
		} else {
			EllipseMotion ellipseMotion = OrbitCalcMotion.calcEllipseOrbit(mass, radius, Hp, Vp);
			results.append("1я космическая:").append(MathUtils.round(ellipseMotion.getV1(), 6)).append("м/с\n");
			results.append("2я космическая:").append(MathUtils.round(ellipseMotion.getV2(), 6)).append("м/с\n");
			results.append("Скорость апогея:").append(MathUtils.round(ellipseMotion.getVa(), 6)).append("м/с\n");
			results.append("Эксцентриситет:").append(ellipseMotion.getEcc()).append("\n");
			results.append("Апогей:").append(MathUtils.round(ellipseMotion.getHa(), 6)).append("\n");
			results.append("Ускорение:").append(MathUtils.round(ellipseMotion.getUskorenie(), 6)).append("м/с2\n");
			int hour = ellipseMotion.getHour();
			int min = ellipseMotion.getMin();
			int sec = ellipseMotion.getSec();
			results.append("Период:").append(DateTimeUtils.convertTime(DateTimeUtils.convertTime(hour, min, sec)));
		}
		tvResult.setText(results.toString());
	}

	private double validateInputDouble(EditText editText){
		String s = editText.getText().toString();
		if (Utility.empty(s)) {
			return 0.0;
		}
		s = s.trim();
		s = Utility.match(s, "-?\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d+)?", 0);
		return Double.parseDouble(s);
	}

	private void validateVariables() {
		mass = validateInputDouble(edtMass);
		radius = validateInputDouble(edtRadius);
		Hp = validateInputDouble(edtHeightPeri);
		Vp = validateInputDouble(edtVelPeri);
	}

	private void initUI(View view) {
		edtMass = view.findViewById(R.id.edtMass);
		edtRadius = view.findViewById(R.id.edtRadius);
		edtHeightPeri = view.findViewById(R.id.edtHeightPeri);
		edtVelPeri = view.findViewById(R.id.edtVelPeri);
		tvResult = view.findViewById(R.id.tvResult);
	}

}

File: app/src/main/java/com/arny/celestiatools/ui/fragments/OrbitCalcFragment.kt
package com.arny.celestiatools.ui.fragments

import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import com.arny.arnylib.utils.CalculatorDialog
import com.arny.arnylib.utils.DateTimeUtils
import com.arny.arnylib.utils.MathUtils
import com.arny.arnylib.utils.Utility
import com.arny.celestiatools.R
import com.arny.celestiatools.utils.astro.AstroConst
import com.arny.celestiatools.utils.astro.OrbitCalc
import kotlinx.android.synthetic.main.fragment_orbit_calc.*

class OrbitCalcFragment : Fragment() {

    private var mass: Double = 0.0
    private var radius: Double = 0.0
    private var Hp: Double = 0.0
    private var Ha: Double = 0.0
    private var Vp: Double = 0.0
    private var ecc: Double = 0.0
    private var sma: Double = 0.0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_orbit_calc, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initListeners()
    }

    private fun initListeners() {
        edtMass.setOnClickListener({ view ->
            CalculatorDialog(context, "Масса", edtMass.text.toString()) { result ->
                edtMass.setText(result)
                calc()
            }.show()
        })
        edtRadius.setOnClickListener({ view ->
            CalculatorDialog(context, "Радиус", edtRadius.text.toString()) { result ->
                edtRadius.setText(result)
                calc()
            }.show()
        })
        edtHeightPeri.setOnClickListener({ view ->
            CalculatorDialog(context, "Перигей", edtHeightPeri.text.toString()) { result ->
                edtVelPeri.setText("")
                edtEcc.setText("")
                edtSma.setText("")
                edtHeightPeri.setText(result)
                calc()
            }.show()
        })
        edtApoHeight.setOnClickListener({ view ->
            CalculatorDialog(context, "Апогей", edtApoHeight.text.toString()) { result ->
                edtVelPeri.setText("")
                edtEcc.setText("")
                edtSma.setText("")
                edtApoHeight.setText(result)
                calc()
            }.show()
        })
        edtVelPeri.setOnClickListener({ view ->
            CalculatorDialog(context, "Скорость перигея", edtVelPeri.text.toString()) { result ->
                edtApoHeight.setText("")
                edtEcc.setText("")
                edtSma.setText("")
                edtVelPeri.setText(result)
                calc()
            }.show()
        })
        edtEcc.setOnClickListener({ view ->
            CalculatorDialog(context, "Эксцентриситет", edtEcc.text.toString()) { result ->
                edtVelPeri.setText("")
                edtApoHeight.setText("")
                edtHeightPeri.setText("")
                edtEcc.setText(result)
                calc()
            }.show()
        })
        edtSma.setOnClickListener({ view ->
            CalculatorDialog(context, "Большая полуось", edtSma.text.toString()) { result ->
                edtVelPeri.setText("")
                edtApoHeight.setText("")
                edtHeightPeri.setText("")
                edtSma.setText(result)
                calc()
            }.show()
        })
        btnFill.setOnClickListener({ view ->
            edtMass.setText(AstroConst.Emass.toString())
            edtRadius.setText(AstroConst.R_Earth.toString())
        })
    }

    private fun calc() {
        validateVariables()
        val results = StringBuilder()
        if (Vp > 0 || Ha > 0 || ecc != 0.0 || sma != 0.0) {
            val ellipseMotion = OrbitCalc.calcEllipseOrbit(mass, radius, Hp, Ha, Vp, ecc, sma)
            results.append("1я космическая:").append(MathUtils.round(ellipseMotion.v1, 3)).append("(м/с)\n")
            results.append("2я космическая:").append(MathUtils.round(ellipseMotion.v2, 3)).append("(м/с)\n")
            results.append("Скорость апогея:").append(MathUtils.round(ellipseMotion.va, 3)).append("(м/с)\n")
            results.append("Скорость перигея:").append(MathUtils.round(ellipseMotion.vp, 3)).append("(м/с)\n")
            results.append("Большая полуось:").append(MathUtils.toExpo(ellipseMotion.sma / 1e3, 6)).append("(км)\n")
            results.append("Эксцентриситет:").append(ellipseMotion.ecc).append("\n")
            edtEcc.setText(ellipseMotion.ecc.toString())
            edtApoHeight.setText(MathUtils.round(ellipseMotion.ha, 3).toString())
            edtHeightPeri.setText(MathUtils.round(ellipseMotion.hp, 3).toString())
            edtVelPeri.setText(MathUtils.round(ellipseMotion.vp, 3).toString())
            edtSma.setText(MathUtils.round(ellipseMotion.sma, 3).toString())
            results.append("Перигей:").append(MathUtils.toExpo(Hp / 1e3, 6)).append("(км) \n")
            results.append("Апогей:").append(MathUtils.toExpo(ellipseMotion.ha / 1e3, 6)).append("(км) \n")
            results.append("Ускорение:").append(MathUtils.round(ellipseMotion.uskorenie, 3)).append("(м/с2)\n")
            val hour = ellipseMotion.hour
            val min = ellipseMotion.min
            val sec = ellipseMotion.sec
            results.append("Период:").append(DateTimeUtils.convertTime(DateTimeUtils.convertTime(hour, min, sec))).append(" : ").append(ellipseMotion.ptime.toLong().toString()).append("(сек)")
        } else {
            val circularMotion = OrbitCalc.calcCircularOrbit(mass, radius, Hp)
            results.append("1я космическая:").append(MathUtils.round(circularMotion.v1, 3)).append("(м/с)\n")
            results.append("2я космическая:").append(MathUtils.round(circularMotion.v2, 3)).append("(м/с)\n")
            results.append("Ускорение:").append(MathUtils.round(circularMotion.uskorenie, 3)).append("(м/с2)\n")
            edtEcc.setText(circularMotion.ecc.toString())
            val hour = circularMotion.hour
            val min = circularMotion.min
            val sec = circularMotion.sec
            results.append("Период:").append(DateTimeUtils.convertTime(DateTimeUtils.convertTime(hour, min, sec))).append(" : ").append(circularMotion.ptime.toLong().toString()).append("(сек)")
        }
        tvResult.text = results.toString()
    }

    private fun validateInputDouble(editText: EditText): Double {
        var s: String? = editText.text.toString()
        if (Utility.empty(s)) {
            return 0.0
        }
        s = s?.trim { it <= ' ' }
        s = Utility.match(s, "-?\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d+)?", 0)
        return java.lang.Double.parseDouble(s)
    }

    private fun validateVariables() {
        mass = validateInputDouble(edtMass)
        radius = validateInputDouble(edtRadius)
        Hp = validateInputDouble(edtHeightPeri)
        Ha = validateInputDouble(edtApoHeight)
        Vp = validateInputDouble(edtVelPeri)
        ecc = validateInputDouble(edtEcc)
        sma = validateInputDouble(edtSma)
    }

}
